### PR TITLE
feat: BSC Nerve + Dodo + Nerve + Ellipsis

### PR DESCRIPTION
**Pull request by dekz**
## Description

* Adds in DODO v1
* Adds Nerve, a modified fork of Curve
* Adds Ellipsis, a direct fork of Curve
```
SELL BUSD->BNB 325496.97 -> 1214.75 ($325496.97) after 132.0s @ bsc
	✔ PASS 325496.97 -> 1214.16 ($303537.23) ($1.91)
	DODO: 5.646%, DODO_V2: 8.503%, PancakeSwap: 85.85000000000001%
	gas: 764657
SELL BUSD->BUSDT 41205.80 -> 41189.97 ($41205.80) after 44.1s @ bsc
	✔ PASS 41205.80 -> 41189.97 ($41189.26) ($0.71)
	DODO: 100%
	gas: 285166
SELL BUSD->BUSDT 86548.13 -> 86513.64 ($86548.13) after 74.9s @ bsc
	✔ PASS 86548.13 -> 86514.45 ($86513.26) ($1.18)
	DODO: 46.089999999999996%, Nerve: 53.910000000000004%
	gas: 472774
SELL BUSD->USDC 20419.59 -> 20410.98 ($20419.59) after 67.3s @ bsc
	✔ PASS 20419.59 -> 20410.98 ($20410.14) ($0.84)
	Nerve: 100%
	gas: 335840
SELL USDC->DAI 408395.37 -> 406804.24 ($408395.37) after 58.7s @ bsc
	✔ PASS 408395.37 -> 406804.24 ($406792.62) ($11.63)
	Belt: 100%
	gas: 4650894
SELL BUSD->USDC 401.65 -> 401.48 ($401.65) after 49.0s @ bsc
	✔ PASS 401.65 -> 401.48 ($400.76) ($0.72)
	Ellipsis: 100%
	gas: 288904
```

```json
{
    "erc20Proxy": "0x95e6f48254609a6ee006f7d493c8e5fb97094cef",
    "exchangeProxy": "0xdef1c0ded9bec7f1a1670819833240f027b25eff",
    "exchangeProxyOwnerStorageSlot": "0x0000000000000000000000000000000300000000000000000000000000000000",
    "taker": "0xd00d00caca000000000000000000000000001337",
    "gst": "0x0000000000b3F879cb30FE243b4Dfee438691c04",
    "transformers": {
        "deployer": "0x8224aa8FE5c9f07d5a59c735386fF6CC6AAEB568",
        "overridesByNonce": {
            "5": {
                "artifactPath": "../0x-monorepo/contracts/zero-ex/test/generated-artifacts/FillQuoteTransformer.json",
                "constructorArgs": [
                    "0xe3a207e4225d459095491ea75d30b31968dff887",
                    "0xdef1c0ded9bec7f1a1670819833240f027b25eff"
                ],
                "balance": 0
            }
        }
    },
    "overrides": {},
    "newFeatures": {
        "deployer": "0xfefefefefefefefefefefefefefefefefefefefe",
        "contracts": []
    },
    "deployments": {
        "deployer": "0xdededededededededededededededededededede",
        "initialNonce": 0,
        "contracts": [
            {
                "artifactPath": "../0x-monorepo/contracts/zero-ex/test/generated-artifacts/BridgeAdapter.json",
                "constructorArgs": [
                    "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c"
                ]
            }
        ]
    }
}
```

TODO
- [ ] Deploy BridgeAdapter + FQT on BSC

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    function _tradeNerve(
+        IERC20TokenV06 sellToken,
+        IERC20TokenV06 buyToken,
+        uint256 sellAmount,
+        bytes memory bridgeData
+    )
+        internal
+        returns (uint256 boughtAmount)
+    {
+        // Basically a Curve fork but the swap option has a deadline
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol:51]: :weary:

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        // Decode the bridge data to get the Curve metadata.
+        NerveBridgeData memory data = abi.decode(bridgeData, (NerveBridgeData));
+        sellToken.approveIfBelow(data.pool, sellAmount);
+        uint256 beforeBalance = buyToken.balanceOf(address(this));
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol:56]: Can we just trust the `dy` returned by `swap()`?

**Review by dorothy-zbornak — APPROVED** [review]
We can deploy a nerve LP VIP if this has good liquidity

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol (hunk 2 above)
Comment (dekz) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinNerve.sol:56]: Oh woops missed that early guy, yeah I just trust the returned value, let me rm this line

**Pull request merged**